### ISSUE TITLE
Submit with a redux action (v5 & v6)

### ISSUE THREAD

**Issue by clayne11** [erikras/redux-form#908]
The docs talk about submitting from a parent, but that requires having a ref to the form. I'm finding myself more and more with multiple containers wrapping around some of my components as HOCs. This means having refs to a form instance is often pretty ugly and hacky and relies on implementation details of the UI which I don't believe it should have to.

There's currently no method for calling `handleSubmit` from a redux action. If there was, you could submit the form from anywhere in the app by simply firing a redux action and you don't have to worry about the implementation of the UI.

An example of where this is very valuable:

I have a form which I use in multiple different locations. In one of the locations, there are "forward / back" buttons at the bottom of the page. The forward button should submit the form (if it is valid) and move on the to the "next" page. If the form is not valid, however, I don't want to move on to the next page.

Currently I don't see a way to do this through redux. Since the "forward / back" buttons aren't part of the form itself, but actually part of the form's container, and since the form might have data fetching containers (or other HOCs) around it, I can't directly access it with a ref to call `submit()`. There's also no way that I can see to tell if the form is valid (or pass down async validation errors) without calling `handleSubmit`.

Using a redux action seems like a much cleaner solution to me than using `ref`.

**Comment by erikras** [comment]
The problem is that the `onSubmit` function lives with the component. So how would dispatching an action through the reducer result in the component submitting?  The `SUBMIT` action would have to set some "submit as soon as you see this" flag in the Redux store, and the component (only one of the potentially many connected form components) would have to respond to it. It gets very messy.

But alternative suggested solutions are welcome.

**Comment by clayne11** [comment]
I think it can be done better than that.

If you fully decouple your UI from your business logic, and all of your validation and submit logic is in the redux reducer, you will have access to all of the validation and submission logic from anywhere in your app where you have access to `dispatch`.

In order to submit, the [`handleSubmit`](https://github.com/erikras/redux-form/blob/master/src/handleSubmit.js#L6) function takes in data and actions from the redux store, as well as a few user provided functions: `validate`, `submit`, and `asyncValidate`.

If we turn `handleSubmit` into a redux action, as long as we provide the form name and the functions `validate`, `submit`, and `asyncValidate` when invoking the action it should function exactly the same way as it does invoking `handleSubmit` from within the form.

I admit that this is a somewhat advanced usage of the library, as it is mentioned in the docs that using any of the exported actions from `redux-form` shouldn't need to be done in simple cases. It does give a great degree of flexibility, though, in how we use the package and in allowing us to decouple business logic from view logic. It also decouples business logic from specific implementations of the view layer (aka using `refs`).

Would you be open to a PR if put one together?

**Comment by erikras** [comment]
Interesting. I was misunderstanding you because of the awkward nuance between "action" and "action creator".

I'm not ready to attempt to implement this at the moment, but I don't see any reason why what you propose would not be possible in a future version.

**Comment by clayne11** [comment]
Would you be open to a PR on v5? I just started using v5 (since v6 is missing some key functionality like array fields) and this is the one missing piece from making this pretty close to a perfect form library in my eyes.

**Comment by fhelwanger** [comment]
I know that you might have other issues to not use v6, but `FieldArray` is available in [alpha.7](https://github.com/erikras/redux-form/releases/tag/v6.0.0-alpha.7)

**Comment by erikras** [comment]
I'd prefer the PR to be in `v6`, but the `handleSubmit` logic is pretty identical between the two, so it probably wouldn't be too hard to port from `v5` to `v6` if you did it for `v5`.

**Comment by clayne11** [comment]
@fhelwanger `FieldArray` is actually all that I needed! I'm going to migrate to `v6` and I'll put in a PR for redux action submission when I have some free time.

**Comment by clayne11** [comment]
I'm starting to have a look at this. Here's what I have so far:

1. The registered fields need to be stored in the redux store, not on the form component. By putting `this.fields` on the form component itself there's no way outside of the form to validate the values. If we register the fields in the redux store, we will always know which fields are registered on the page.

  Does this cause any issues with any intended use cases? The only issue I can think of is having multiple of the same form with different fields mounted at the same time. In this case, we won't know which fields are mounted to which instance of the form. I know the wizard docs suggest using the same form name with different mounted fields, but I assume only one of them would be mounted at a time so that shouldn't be an issue.

2. Valid is currently calculated by asking each form field whether or not they are valid. This doesn't make very much sense to me, though, since the values and all of the validation errors are available at a higher abstraction. 

  To determine validity of the form, we need:
  * list of registered fields
  * async errors
  * submit errors
  * form values and validate function OR sync errors
  
  Now that we have solved my first point, and the list of registered fields is in the redux store, we can validate the form by moving the field validation logic out of the field and into a common utility function that takes in the parameters listed above, which gives us the ability to validate the form outside of the form context by simply passing in the validation function since the rest of the data comes out of the redux store.

  Another option is to calculate the sync errors any time the form is changed and store those in the redux store as well alongside the async and submit errors, rather than calculating sync errors during the render. I like this solution because we now no longer even need access to the validation function to validate the form when we are outside of the form context. We can essentially run a reducer on the form reducer to determine if a form is valid. This seems like much more idiomatic `redux` to me.

In summary, I think we should:

1. Store the registered field for a given form in the redux store
1. Calculate and store the sync errors in the redux store whenever a field is changed rather than calculating the errors at render time

What do you think?

**Comment by clayne11** [comment]
I want to accommodate future use cases like https://github.com/erikras/redux-form/issues/917, so I think we should store the form validation function alongside the field name when we register a new field.

Here is my proposal for the store state:

```javascript
// redux store
{
	form: {
		myForm: {
			registeredFields: {
				foo: {
					validate: <optional function>
				}
			},
			values: {
				foo: 'bar'
			},
			initial: {
			    foo: 'baz'
			},
			syncErrors: {
			    foo: 'bar is not a valid value'
			}
		}
	}
}
```

If you think this looks good I'm going to move forward with it.

**Comment by erikras** [comment]
Interesting analysis, @clayne11. You have understood the code very well. I have toyed with each of these ideas in the past. [`v6.0.0-alpha.4` kept the sync errors in the store](http://redux-form.com/6.0.0-alpha.4/examples/syncValidation/).

1. So every field dispatches a `REGISTER` and `UNREGISTER` action on mount and unmount respectively? This will annoy the people who complain that `redux-form` clogs up their redux-devtools, but they're wrong. I approve of this change.

2. > Valid is currently calculated by asking each form field whether or not they are valid. This doesn't make very much sense to me, though, since the values and all of the validation errors are available at a higher abstraction.

Yep, this was done in an attempt to only validate the fields that were rendered, but with # 1, that reason is no longer valid (ha!). However, in the future, I would like to provide the ability to provide sync validation as a prop to `<Field>`, in which case the outer form component _must_ query all its field components.

> Another option is to calculate the sync errors any time the form is changed and store those in the redux store as well alongside the async and submit errors, rather than calculating sync errors during the render. I like this solution because we now no longer even need access to the validation function to validate the form when we are outside of the form context. We can essentially run a reducer on the form reducer to determine if a form is valid. This seems like much more idiomatic redux to me.

As I mentioned above, this was my original intent of `v6`, but I got quite a few complaints from people who were passing in a custom validation function for each instance of their form component, and just general complaints about making the `v5`-to-`v6` migration distance longer. However, there is no reason why a hybrid approach could not work, where you _could_, if you chose to, provide your validation function at the reducer level, but also allow the validation as a prop/config param.

A potential middle ground that I have considered would be for each field to also submit the new value for `syncError` on every `CHANGE` action. With the current API, this would require an `onChange` function that looked sort of like this:

```js
onChange = event => {
  const value = getValue(event)
  const newAllFormValues = setIn(allFormValues, name, value)
  const allErrors = formWideValidation(newAllFormValues)
  const error = getIn(allErrors, name)
  dispatch(change(name, value, error))
}
```

That solves all of our requirements of having the sync errors in redux, but the validation function potentially just in the form component or, in the future, in the field component.

>  I think we should store the form validation function alongside the field name when we register a new field.

It's awfully tempting, but putting functions in your store is an antipattern. For proper atomic playback, actions can be data only.

**Comment by clayne11** [comment]
Good point about the playback. I forgot about that `redux` requirement.

1. That's exactly it. Each form component will `REGISTER` and `UNREGISTER` itself on mount and unmount. Essentially where you currently have `register` on the form store the state on itself, it will instead just emit an action to put that state into the store.

I completely agree with having each field return its own sync error if it was given a validate function.

The key in making all of this work and being not only easier to reason about but more flexible is getting all of the state into the store rather than having certain data like the sync errors calculated on render.

By allowing a `Field` to optionally have a `validate` function and return its own sync errors in that case it pushes all of the form state up into the store and allows us to them submit from a redux action.

Your `onChange` is missing the `Field` specific validate function though, right? So it should look more like this if we move to allowing a `validate` prop on a field:

```javascript
onChange = (event, validate) => {
  const value = getValue(event)
  const newAllFormValues = setIn(allFormValues, name, value)
  const allErrors = formWideValidation(newAllFormValues)
  const error = getIn(allErrors, name) || validate && validate(value)
  dispatch(change(name, value, error))
}
```

So looping back on the state proposal:
```
{
    form: {
        myForm: {
            registeredFields: {
                foo: {}
            },
            values: {
                foo: 'bar'
            },
            initial: {
                foo: 'baz'
            },
            syncErrors: {
                foo: 'bar is not a valid value'
            }
        }
    }
}
```

**Comment by clayne11** [comment]
I took a crack and moving all of the field registration into the redux store. I also pushed up the validation logic into he form at render time alongside `syncError` calculation.

I'm going to look at moving the `syncError` into the store next.

**Comment by erikras** [comment]
> Your `onChange` is missing the `Field` specific validate function though, right?

Yes, that's why I said "With the current API...". 😄 

> the `deleteIn` function was trying to look into the path I provided rather than treat it as a key

Haha! I guess your options are:

1. use an array
2. replace every `.` with some other character that we hope people won't use in their field names, like `~`
3. write a non-recursing version of `deleteIn`

The third is probably best.

> I'm curious, since you had the `syncError` in the store in `alpha.4`, why did you later move it out?

Because it didn't need to be there, and I generally agree with the complaints that I was getting that it _feels_ more like the validation should be closer to the component, so you when you go to add a new field, it's all in the same spot. And one of the things I like about `redux-form` is that, for the most part, you just have to set up your reducer once, and then forget about it, no matter how many forms you create. For most usage, you don't care about the reducer at all.

**Comment by clayne11** [comment]
I disagree. Putting the `syncErrors` in the store is more flexible because it moves what is essentially a hidden piece of state in the `syncErrors` into the store and allows for things like submitting with an action.

In simple cases I agree that you don't need to worry about the reducer, but it's always nice to be more flexible.

I'm trying to figure out the best way to do the sync validation now. How would you access all the form values as you've shown in your gist? I spent some time looking at it but I couldn't think of a good way of doing it that wouldn't incur a lot more re-renders or require passing a function into the reducer which I'm thinking is an anti-pattern.

**Comment by erikras** [comment]
> Putting the syncErrors in the store is more flexible because it moves what is essentially a hidden piece of state in the syncErrors into the store and allows for things like submitting with an action.

To be clear, I agree that we should put sync errors in the store. I used to be against it, but I've changed my mind. Hence that we're even having this discussion. 😄 

>  How would you access all the form values as you've shown in your gist?

The same tricky way that we are currently accessing all of the sync errors. This was a pretty major breakthrough (for me, at least) in `v6`, realizing that the outer form component can have access to all the sync errors and values without rerendering every time they change, and then a function to query them can be passed down to the fields when they register themselves.

**Comment by clayne11** [comment]
Can you have a look at my PR for the field registration when you get a minute? From my end I think it's ready to go so if you have feedback I'd like to get that going. It can be pulled in on it's own and moving the `syncErrors` into the store will be a separate PR.

I don't mean to bug you but I do see that you've been quite active today so I was just hoping we could move this along.

**Comment by erikras** [comment]
Any further progress on this, @clayne11?

**Comment by clayne11** [comment]
I haven't had a chance to work on it in the last week or so since I pushed the PR for registering fields. I'll take another look and try to get it done this weekend.

**Comment by clayne11** [comment]
This is proving to be more work than I initially anticipated.

I have moved the sync errors into the store, however I still haven't handled all of the array cases (shift, push, insert, pop, etc).

Since sync validation originally happened on every render, the validation code only had to run in one place - inside the form connect function.

Now that the errors are being passed up by the fields, though, the validation needs to run anytime an action changes the form values.

I have added sync validation to `onChange` as well as `registerField` so that the validation errors are updated anytime a field is registered or a field is changed.

The validation needs to also run anytime an array field changes, though, which is much more complicated than running during the `onChange` event.

In order to run sync validation we must first update the value that is going to be pushed to the store before running the validation.

`onChange` looks like this:

```javascript
const createOnChange = (
  change,
  syncValidate,
  getAllValuesAndProps,
  name,
  structure) =>
  event => {
    const value = getValue(event, isReactNative)
    const syncErrors = getSyncErrors({
      value, getAllValuesAndProps, name, syncValidate
    }, structure)
    change(value, syncErrors)
  }
```

```javascript
// getSyncErrors.js
export default ({
  name, value, syncValidate, getAllValuesAndProps
}, { setIn }) => {
  const { allValues, props } = getAllValuesAndProps()
  const newAllValues = setIn(allValues, name, value)
  return syncValidate && syncValidate(newAllValues, props)
}
```

This is simple because updating `allValues` simply requires running `setIn(allValues, name, value)`.

With arrays, however, the transformation is much more expensive and complex.

```javascript
  const doSplice = (state, key, field, index, removeNum, value, force) => {
    const existing = getIn(state, `${key}.${field}`)
    return existing || force ?
      setIn(state, `${key}.${field}`, splice(existing, index, removeNum, value)) :
      state
  }
  const rootKeys = [ 'values', 'fields', 'submitErrors', 'syncErrors', 'asyncErrors' ]
  const arraySplice = (state, field, index, removeNum, value, syncErrors) => {
    let result = state
    result = doSplice(result, 'values', field, index, removeNum, value, true)
    result = doSplice(result, 'fields', field, index, removeNum, empty)
    result = doSplice(result, 'submitErrors', field, index, removeNum, empty)
    result = doSplice(result, 'asyncErrors', field, index, removeNum, empty)
    result = updateSyncErrors(result, syncErrors, structure)
    return result
  }
```

In order to calculate the sync errors before passing them to the relevant action we must first perform the array transformation so that the sync validation runs on the updated value.

I could factor out the `arraySplice` and all of the array transformation functions into separate files that could be imported and run whenever an action like `arrayInsert` is run. Going this route, though, we're going to end up performing the array transformations twice: once to run the sync validation on the updated value and once in the reducer to update the value in the store.

We could always just pass the updated store value from the action itself but that somewhat defeats the purpose of the action.

How do you think we should go about this?

**Comment by erikras** [comment]
Yep, sounds like you're on exactly the right track. Great work on this!

You've got to perform the "set" (or "splice") outside of the reducer in order to run the validation what the values _will be_ outside of the reducer. Therefore, it's more efficient to do the validation _inside_ the reducer. _But_, some people like the API where the validation is outside, so we've got to implement the both the efficient way (inside) and the inefficient way (outside). If someone complains that outside-the-reducer validation is slowing down their form, we tell them to move it to the reducer.

**Comment by clayne11** [comment]
Quick update on this: I was quite sick this past weekend and I also ran into some issues with my (paid) work so I lost a few days to work on this. I'm not going to be able to look at it again until next weekend.

I hope to have the `syncErrors` fully moved into the store by the end of next weekend which I think is 80-90% of the work required to perform submission with a Redux action.

**Comment by diegohaz** [comment]
Any updates on this? It would be really useful to have a `submit` action creator. I'm currently racking my brains to properly submit a form from outside the component.

**Comment by diegohaz** [comment]
Hey, guys, I've published [redux-form-submit](https://github.com/diegohaz/redux-form-submit), which adds the submit action creator to redux-form. Maybe we can add this to redux-form itself in the future.

**Comment by yanndinendal** [comment]
> I'm currently racking my brains to properly submit a form from outside the component.

Did you try something like this?

``` html
<button form="my-form-id">Submit!</button>
<form id="my-form-id">
  <input value="Hello World!" />
</form>
```

**Comment by diegohaz** [comment]
@yanndinendal Actually, I need to imperatively submit the form. But I didn't know that `form` attribute. Thanks. :)

**Comment by melbayad** [comment]
@yanndinendal In my case, I have the submit button in my header app component and form in other one.  I've tried your suggested solution but it doesn't work. I posted this [issue](http://stackoverflow.com/questions/40513513/redux-form-calling-submit-outside-form-component)... any help will be welcome.

**Comment by gustavohenke** [comment]
Dudes, this has been possible for quite some time now!
Please check the [remote submit example](http://redux-form.com/6.4.3/examples/remoteSubmit/).
